### PR TITLE
Add support for strings and __future__ annotations

### PR DESCRIPTION
**Pull request by dmontagu**
This pull request adds support for forward references, string annotations, and `from __future__ import annotations` in decorated endpoint signatures.

### CODE REVIEW

File: fastapi/dependencies/utils.py [hunk 1]
```diff
@@ -158,6 +160,36 @@ def is_scalar_sequence_field(field: Field) -> bool:
...
+        # elif hasattr(type(call), "__call__"):
+        #     type_hints = get_type_hints(type(call).__call__)
+        # else:
+        #     raise ValueError("Could not determine callable call")
+    # elif hasattr(call, "__call__"):
+    else:
+        type_hints = get_type_hints(call.__call__)  # type: ignore
+    # else:
+    #     raise ValueError("Could not determine callable call")
+    typed_params = []
```
Comment (dmontagu) [inline fastapi/dependencies/utils.py]: I commented out a variety of edge-case checks here to ensure code coverage would pass, but before merging this code should be uncommented with appropriate tests added.

**Comment by tiangolo** [comment]
Nice! Good job!

> This PR would introduce a small breaking change [...]

No problem with that. We bump the version to release it.

> Before this is ready to merge [...]

Perfect. Let me know when you can check those things.

**Comment by dmontagu** [comment]
@tiangolo Thanks for the feedback!

I actually discussed this with Ivan Levkivskyi on the python/typing gitter, and he suggested I make a bpo post (which I did here: https://bugs.python.org/issue37923). So far it hasn't gotten much attention 😕 but I figured I'd at least ask the question.

Having dug a little more into the various related PEPs and python stdlib source, I feel pretty good about the approach in this PR, so I'm inclined to run with it, and just add more edge-case handling and tests for them. Hopefully in the future there will be first class support for this built directly into the `inspect` or `typing` packages, and we can remove any custom/redundant logic.

I'll try to get this into a mergeable state this weekend.

File: tests/test_security_oauth2.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (dmontagu) [inline tests/test_security_oauth2.py]: This file looked like a good example of complex dependency usage, and allowed me to get 100% coverage for tests.

I also wanted to start by making a minimal set of changes to make the pull request easier to review; simply adding `from __future__ import annotations` seemed like an easier way to prove it works than creating an entirely new test file to review.

@tiangolo How would you propose we handle testing this functionality?
* Should I copy this file (or another) and just add the `from __future__ import annotations` import?
* Should I create an entirely new file with a more comprehensive set of string-annotation tests?
    * In this case, is there a good place for me to look for examples?

File: fastapi/dependencies/utils.py [hunk 3]
```diff
@@ -171,6 +171,30 @@ def is_scalar_sequence_field(field: Field) -> bool:
     return False
 
 
+def get_typed_signature(call: Callable) -> inspect.Signature:
```
Comment (dmontagu) [inline fastapi/dependencies/utils.py]: I originally wanted to use `typing.get_type_hints` directly, but ultimately this implementation seemed cleaner and follows `typing.get_type_hints` where it matters. It also uses `pydantic.utils` to handle differences between the `typing` api for python 3.6 and 3.7 (for `ForwardRef` and `evaluate_forwardref` anyway), and to deal with the fact that the relevant objects are technically private in `typing`. Since `pydantic` has made them public, it seems reasonable to let `pydantic` worry about ensuring the public API is stable.

**Comment by dmontagu** [comment]
@tiangolo I reworked the implementation, and now it should be 100% backward compatible (no changes to the annotations in the case of optionals, etc.), and only change any of the behavior when string-valued annotations are present.

**Comment by dmontagu** [comment]
Note: the currently-failing test is due to the `from __future__ import annotations` line I added; I'll remove it shortly.

File: tests/test_security_oauth2.py [hunk 4]
```diff
@@ -21,18 +21,18 @@ class User(BaseModel):
     username: str
 
 
-def get_current_user(oauth_header: str = Security(reusable_oauth2)):
+def get_current_user(oauth_header: "str" = Security(reusable_oauth2)):
```
Comment (dmontagu) [inline tests/test_security_oauth2.py]: It might be better to make this change in another file; @tiangolo if you give me some direction I can implement the tests however you'd prefer.

**Review by tiangolo — CHANGES_REQUESTED** [review]
Awesome job! :tada: :rocket: 

I think it's fine to leave the tests with your modifications and test both things at the same time. I just requested comments on top of them to remind us to *not* "refactor" them into their non-string version.

**Comment by dmontagu** [comment]
@tiangolo

* I made the requested changes.
* I also updated the `# type: ignore` comments to:
    1. be compatible with the recent mypy 0.730 release (which is automatically used in the build)
    2. remove unused ignores (detected by calling `mypy fastapi --disallow-untyped-defs --warn-unused-ignore`)
* Do you think the docs should be updated anywhere? It seems like release notes is probably enough

File: fastapi/utils.py [hunk 5]
```diff
@@ -58,10 +58,10 @@ def create_cloned_field(field: Field) -> Field:
     use_type = original_type
     if lenient_issubclass(original_type, BaseModel):
         original_type = cast(Type[BaseModel], original_type)
-        use_type = create_model(  # type: ignore
+        use_type = create_model(
             original_type.__name__,
             __config__=original_type.__config__,
-            __validators__=original_type.__validators__,
+            __validators__=original_type.__validators__,  # type: ignore
```
Comment (dmontagu) [inline fastapi/utils.py]: New mypy is smarter about the where the # type: ignore should go

File: fastapi/openapi/models.py [hunk 6]
```diff
@@ -11,7 +11,7 @@
     import email_validator
 
     assert email_validator  # make autoflake ignore the unused import
-    from pydantic.types import EmailStr  # type: ignore
+    from pydantic.types import EmailStr
```
Comment (dmontagu) [inline fastapi/openapi/models.py]: This `# type: ignore` is not necessary for mypy to pass; let me know if you want this in a separate PR

File: fastapi/dependencies/utils.py [hunk 7]
```diff
@@ -405,7 +433,7 @@ def is_coroutine_callable(call: Callable) -> bool:
     values.update(cookie_values)
     errors += path_errors + query_errors + header_errors + cookie_errors
     if dependant.body_params:
-        body_values, body_errors = await request_body_to_args(  # type: ignore # body_params checked above
```
Comment (dmontagu) [inline fastapi/dependencies/utils.py]: This `# type: ignore` is not necessary for mypy to pass; let me know if you want this in a separate PR

File: fastapi/dependencies/utils.py [hunk 8]
```diff
@@ -329,8 +353,12 @@ def is_coroutine_callable(call: Callable) -> bool:
 ]:
     values: Dict[str, Any] = {}
     errors: List[ErrorWrapper] = []
-    response = response or Response(  # type: ignore
-        content=None, status_code=None, headers=None, media_type=None, background=None
+    response = response or Response(
+        content=None,
+        status_code=None,  # type: ignore
```
Comment (dmontagu) [inline fastapi/dependencies/utils.py]: Needed to put the `# type: ignore` here to get mypy to pass (sadly this forces the line to become longer, but has the benefit of *only* applying to that one argument.)

**Review by tiangolo — APPROVED** [review]
Awesome! Thanks a lot!

LGTM :tada: :rocket:
